lib/rubygems.rb

```diff
@@ -773,6 +773,10 @@ def self.write_binary(path, data)
     open_file(path, 'wb') do |io|
       io.write data
     end
+  rescue Errno::ENOSPC => e
+    # If the write failed, delete the file -- it's *guaranteed* to be corrupted.
+    File.delete(path) if File.exist?(path)
```

Comment:
I think an `Errno::ENOSPC` error would _always_ leave a corrupted file so the `File.exist?` check is not necessary?

Reply:
While poking around with this bug, I encountered exactly two scenarios:

1. If you have space left, but not enough for the whole file, it will be corrupt.
2. If you have perfectly filled your hard drive, it will fail with `Errno::ENOSPC` but not create a file.

The `File.delete(path)` addresses # 1 and the `if File.exist?(path)` conditional addresses # 2.

Reply:
I think case 2 is weird, what raises `Errno::ENOSPC`? I assume it's not `File.open` because otherwise it would leave an empty file, but I'm not sure what it is then. Ideally I'd like to cover case 2 with a spec too, that's why I'm asking. Can you reproduce scenario 2 with `/dev/full?

Reply:
My understanding is:

- in scenario 1, the error is raised by `io.write`
- in scenario 2, the error is raised by `File.open`

If `File.open` successfully creates the file, there was enough space. And then `io.write` can fail, giving a corrupt file.

If `File.open` _does not_ successfully create the file, it'd raise some error &mdash; and `Errno::ENOSPC` would make sense. This would result in an error, but no file.

Reply:
When I work on this PR next, I'll try to reproduce.